core/domain/stats_jobs_one_off.py

```diff
@@ -127,3 +143,2369 @@ def reduce(key, stringified_values):
+                })
+
+    @staticmethod
+    def reduce(key, stringified_values):
+        value_dict_list = [
+            ast.literal_eval(stringified_value)
+            for stringified_value in stringified_values]
+
+        # The first dict can be used to extract item_id and rule_name since they
+        # will be the same for all results managed by this call to reduce().
+        # Note that the item_id is only available for non-aggregation results.
+        first_value_dict = value_dict_list[0]
+
+        old_total_count = 0
+        new_total_count = 0
+        for value_str in stringified_values:
+            value_dict = ast.literal_eval(value_str)
+            if value_dict['type'] == (
+                    ClearUnknownMissingAnswersJob._OLD_ANSWER_MODEL_TYPE):
+                old_total_count = old_total_count + value_dict['frequency']
+            else:
+                new_total_count = new_total_count + value_dict['frequency']
+
+        # Only output the exploration ID if the counts differ to avoid
+        # large amounts of output.
+        if old_total_count != new_total_count:
+            item_id = first_value_dict['item_id']
+            migrated_answer_record = stats_models.MigratedAnswerModel.get(
+                item_id, strict=False)
+            if migrated_answer_record:
+                for exp_version in migrated_answer_record.exploration_versions:
+                    all_models = stats_models.StateAnswersModel.get_all_models(
+                        migrated_answer_record.exploration_id, exp_version,
+                        migrated_answer_record.state_name)
+                    if not all_models:
+                        continue
+                    # The entire state of records need to be deleted since there
+                    # isn't a reliable way to go into submitted answers and
+                    # remove specific failures. The entire state will be
+                    # re-migrated the next time the AnswerMigrationJob runs.
+                    for state_answer_model in all_models:
+                        submitted_answers = [
+                            stats_domain.SubmittedAnswer.from_dict(
+                                submitted_answer_dict)
+                            for submitted_answer_dict
+                            in state_answer_model.submitted_answer_list]
+                        exp_id = state_answer_model.exploration_id
+                        state_name = state_answer_model.state_name
+                        for submited_answer in submitted_answers:
+                            rule_str = submited_answer.rule_spec_str
+                            old_item_id = u'%s.%s.%s.%s' % (
+                                exp_id, state_name, 'submit', rule_str)
+                            separate_migrated_answer_record = (
+                                stats_models.MigratedAnswerModel.get(
+                                    old_item_id, strict=False))
```

Comment:
Records are stored based on the old storage model item IDs, which means it's possible for multiple records to correspond to a single state answer model entry (think multiple exploration versions). The old storage model IDs did not contain exploration version, so this isn't a 1:1 mapping and additional cleanup is required for each answer we want to remove.